sigal/plugins/compress_assets.py

```diff
@@ -0,0 +1,153 @@
+        '''
+        Perform actual compression.
+        This should be implemented by subclasses.
+        '''
+        raise NotImplementedError
+
+    def compress(self, filename):
+        '''
+        Compress a file, only if needed.
+        '''
+        compressed_filename = self.get_compressed_filename(filename)
+        if not compressed_filename:
+            return
+
+        self.do_compress(filename, compressed_filename)
+
+    def get_compressed_filename(self, filename):
+        '''
+        If the given filename should be compressed, returns the compressed filename.
+        A file can be compressed if:
+            - It is a whitelisted extension
+            - The compressed file does not exist
+            - The compressed file exists by is older than the file itself
+        Otherwise, it returns False.
+        '''
+        if not os.path.splitext(filename)[1][1:] in self.suffixes_to_compress:
+            return False
+
+        file_stats = None
+        compressed_stats = None
+        compressed_filename = u'{}.{}'.format(filename, self.suffix)
+        try:
+            file_stats = os.stat(filename)
+            compressed_stats = os.stat(compressed_filename)
+        except OSError:  # FileNotFoundError is for Python3 only
+            pass
+
+        if file_stats and compressed_stats:
+            return compressed_filename if file_stats.st_mtime > compressed_stats.st_mtime else False
+        else:
+            return compressed_filename
+
+
+class GZipCompressor(BaseCompressor):
+    suffix = 'gz'
+
+    def do_compress(self, filename, compressed_filename):
+        with open(filename, 'rb') as f_in, gzip.open(compressed_filename, 'wb') as f_out:
+            shutil.copyfileobj(f_in, f_out)
+
+
+class ZopfliCompressor(BaseCompressor):
+    suffix = 'gz'
+
+    def do_compress(self, filename, compressed_filename):
+        import zopfli.gzip
+        with open(filename, 'rb') as f_in, open(compressed_filename, 'wb') as f_out:
+            f_out.write(zopfli.gzip.compress(f_in.read()))
+
+
+class BrotliCompressor(BaseCompressor):
+    suffix = 'br'
+
+    def do_compress(self, filename, compressed_filename):
+        import brotli
+        with open(filename, 'rb') as f_in, open(compressed_filename, 'wb') as f_out:
+            f_out.write(brotli.compress(f_in.read(), mode=brotli.MODE_TEXT))
+
+
+def get_compressor(settings):
+    name = settings.get('method', DEFAULT_SETTINGS['method'])
+    if name == 'gzip':
+        return GZipCompressor(settings)
+    elif name == 'zopfli':
+        try:
+            import zopfli.gzip
+            return ZopfliCompressor(settings)
+        except ImportError:
+            logging.warning('Zopfli not found, using standard gzip')
+            return GZipCompressor(settings)
+
+    elif name == 'brotli':
+        try:
+            import brotli
+            return BrotliCompressor(settings)
+        except ImportError:
+            logger.error('Unable to import brotli module')
```

Comment:
Why no `GZipCompressor` in this case ?

Reply:
Actually I think I prefer this, and it should do the same for zopfli. If a specific (non default) compressor is specified in the settings but the package is not found, it should always raise an error instead of using another compressor.

Reply:
My thought was that zopfli is a (better) gzip compressor. So if a user wanted to use zopfli, but the module is missing, the only drawback is that files might be bigger, but everything will work as expected.
It is not the case for brotli: if the user configured its web server to serve pre-compressed brotli content, gzip files won't match.
I think that a warning is enough for zopfli, but if you insist, I can make it a real error ;)

Reply:
Yes I think that raising an error is better, to have a consistent behavior between options and avoid confusion for a user that will need to find why it doesn't work as expected.